File: 238-Product of Array Except Self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """

# O(n)
        tmp = 1
        res = []
        for i in range(len(nums)):
            res.append(tmp)
            tmp = tmp*nums[i]
        
        tmp2 = 1
        for j in range(len(nums)-1,-1,-1):
            res[j] = res[j]*tmp2
            tmp2 = tmp2*nums[j]        
        return res

File: C++/238-Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int N = nums.size();
        vector<int> left(N);
        vector<int> res(N);
        int curr = 1;
        for (int i=0; i<N; i++) {
            curr *= nums[i];
            left[i] = curr;
        }
        
        curr = 1;
        for (int i= N-1; i>=0; i--) {
            if (i-1>=0) {
                res[i] = curr*left[i-1];
            } else {
                res[i] = 1*curr;
            }
            curr = curr*nums[i];
        }
        
        return res;
    }
};